java/arcs/core/storage/StorageProxy.kt

```diff
@@ -0,0 +1,230 @@
+) {
+    private val mutex = reentrantLock()
+    private val handles: MutableSet<Handle<Data, Op, T>> = mutableSetOf()
+    private val crdt = initialCrdt
+    private val waitingSyncs: MutableList<CompletableDeferred<ValueAndVersion<T>>> = mutableListOf()
+    private var synchronized = false
+    private var listenerAttached = false
+    private var hasReaders = false
+
+    private val store = storeEndpointProvider.getStorageEndpoint()
+
+    /**
+     * Connects a handle. If the handle is readable, it will receive the configured callbacks.
+     */
+    suspend fun registerHandle(handle: Handle<Data, Op, T>): VersionMap {
+        // non-readers don't get callbacks, return early
+        if (!handle.canRead) {
+            mutex.withLock {
+                return crdt.versionMap.copy()
+            }
+        }
+
+        mutex.withLock() {
+            if (!listenerAttached) {
+                store.setCallback(ProxyCallback<Data, Op, T>(::onMessage))
+                listenerAttached = true
+            }
+
+            handles.add(handle)
+
+            // Change to synchronized mode as soon as we get any read handles and send a request to get
+            // the full model (once).
+            if (!this.hasReaders) {
+                this.requestSynchronization()
+                this.hasReaders = true
+            }
+
+            // If a handle registers after we've received the full model, notify it immediately.
+            if (this.synchronized) {
+                coroutineScope {
+                    launch { handle.callback?.onSync() }
+                }
+            }
+            return crdt.versionMap.copy()
+        }
+    }
+
+    /**
+     * Disconnects a handle so it no longer receives callbacks.
+     */
+    suspend fun deregisterHandle(handle: Handle<Data, Op, T>) {
+        mutex.withLock {
+            handles.remove(handle)
+        }
+    }
+
+    /**
+     * Apply a CRDT operation to the [CrdtModel] that this [StorageProxy] manages, notifies read
+     * handles, and forwards the write to the [Store].
+     */
+    suspend fun applyOp(op: Op): Boolean {
+        mutex.withLock {
+            if (!crdt.applyOperation(op)) {
+                return false
+            }
+            val msg = ProxyMessage.Operations<Data, Op, T>(listOf(op), null)
+            store.onProxyMessage(msg)
+            notifyUpdate(op)
+            return true
+        }
+    }
+
+    /**
+     * Return the current local version of the model, as well as the current associated version map
+     * for the data. Suspends until it has a synchronized view of the data.
+     */
+    suspend fun getParticleView(): ValueAndVersion<T> {
+        return getParticleViewAsync().await()
+    }
+
+    suspend fun getParticleViewAsync(): CompletableDeferred<ValueAndVersion<T>> {
+        val future = CompletableDeferred<ValueAndVersion<T>>()
+        mutex.withLock {
+            if (synchronized) {
+                return future.apply {
```

Comment:
Optional Nit: I typically prefer a single `return` at the bottom whenever possible, rather than short-circuiting early (unless the short-circuiting is due to a guard). It makes things slightly easier to read IMO.